sdk/search/search-documents/src/indexModels.ts

```diff
@@ -685,3 +689,105 @@ export type IndexDocumentsAction<T> = {
+
+// Utility types
+
+export type UnionToIntersection<U> =
+  // Distribute members of U into parameter position of a union of functions
+  (
+    U extends unknown ? (_: U) => unknown : never
+  ) extends // Infer the intersection of the members of U as a single intersected parameter type
+  (_: infer I) => unknown
+    ? I
+    : never;
+
+// Types that should not be included in SelectPaths recursion
+export type ExcludedODataTypes = Date | GeographyPoint;
+
+/**
+ * Produces a union of valid Cognitive Search OData $select paths for T
+ * using a post-order traversal of the field tree rooted at T.
+ */
+export type SelectFields<T extends object> = T extends Array<infer U>
+  ? // Allow selecting fields only from elements which are objects
+    NonNullable<U> extends object
+    ? SelectFields<NonNullable<U>>
+    : never
+  : {
+      // Only consider string keys
+      [K in Exclude<keyof T, symbol | number>]: NonNullable<T[K]> extends object
+        ? NonNullable<T[K]> extends ExcludedODataTypes
+          ? // Excluded, so don't recur
+            K
+          : SelectFields<NonNullable<T[K]>> extends infer NextPaths
+          ? // Narrow NextPaths' type for template literal
+            NextPaths extends string
+            ? // Union this key with all the next paths separated with '/'
+              K | `${K}/${NextPaths}`
+            : // We didn't infer any nested paths, so just use this key
+              K
+          : never
+        : // Not an object, so can't recur
+          K;
+    }[Exclude<keyof T, symbol | number>];
+
+/**
+ * Deeply pick fields of T using valid Cognitive Search OData $select
+ * paths.
+ */
+export type SearchPick<T extends object, Paths extends SelectFields<T>> =
+  // We're going to get a union of individual interfaces for each field in T that's selected, so convert that to an intersection.
+  UnionToIntersection<
+    // Paths is a union or single string type, so if it's a union it will be _distributed_ over this conditional.
+    // Fortunately, template literal types are not greedy, so we can infer the field name easily.
+    Paths extends `${infer FieldName}/${infer RestPaths}`
+      ? // Symbols and numbers are invalid types for field names
+        FieldName extends Exclude<keyof T, symbol | number>
+        ? NonNullable<T[FieldName]> extends Array<infer U>
+          ? U extends object
+            ? // Extends clause is necessary to refine the constraint of RestPaths
+              RestPaths extends SelectFields<U>
+              ? // Narrow the type of every element in the array
+                {
+                  [K in FieldName]:
+                    | Array<SearchPick<U, RestPaths>>
+                    | Extract<T[K], null | undefined>;
+                }
+              : // Unreachable by construction
+                never
+            : // Don't recur on arrays of non-object types
+              never
+          : NonNullable<T[FieldName]> extends object
+          ? // Recur :)
```

Comment:
This is another annoying thing, but technically there is a difference between an optional field and a field that might be undefined, and that difference is observable if the user has `exactOptionalPropertyTypes` set in their tsconfig. Based on my testing with this proposal, all optional fields such as `foo?: string` are reduced to unions with undefined, like `foo: string | undefined`.

@dgetu I made an ugly modification here. What do you think? `exactOptionalPropertyTypes` is a pretty niche feature, and since this is an output refinement it might not actually cause any problems if we don't fix it. FWIW I think it's annoying to have to duplicate all this but maybe you'll have some good ideas about ways it could be cleaned up?

```suggestion
          ? // Recur :)
            undefined extends T[FieldName]
            ? {
                // Since `undefined` is a possible value, we'll create an optional field. `undefined` cannot be serialized in JSON, so it really will be missing at runtime.
                [K in FieldName]?: RestPaths extends SelectFields<
                  T[K] & {
                    // This empty intersection fixes `NonNullable<T[K]>` not being narrowed to an object type in older versions of TS
                  }
                >
                  ?
                      | SearchPick<
                          T[K] & {
                            // Ditto
                          },
                          RestPaths
                        >
                      | (null extends T[K] ? null : never)
                  : // Unreachable by construction
                    never;
              }
            : 
```

Reply:
I think the answer to this is a lot more complicated than it look. I'm not even sure if optional fields makes sense here. From my previous experience with the Azure Search service (though I would appreciate if someone could validate this), it is impossible to get an `undefined` value from the search query. When inserting documents into an index, if some of the expected fields (based on the schema of the index) are missing, they will be indexed as `null`. When querying the index, you will get this `null` value back. Because of this, every field in the type can potentially be null, but never missing unless the type includes more fields that are present in the schema (in which case they will always be missing).

However, it is possible that the client itself changes this behavior to erase null fields from the returned documents.

Reply:
That's an interesting TS feature. I'm testing this change with TS 4.4 (the version where `exactOptionalPropertyTypes` was introduced), and this change is yielding `never` for my test type. I'll dig into this.

Reply:
It sounds like there may not be any need to handle optional since the values may come back null instead of omitted unless we explicitly `select` some other fields, but if so we should at least update the `Hotel` example interface so that it uses exact types supported well by the service.

Reply:
Having looked at the deserialization logic, it looks like the response from the service and the deserializer won't ever return `undefined`. This being the case, it's probably a good idea to support clients who provide wider-than-necessary model types, but I don't think we need to support generating a similarly wide type to match.

Reply:
@dgetu This means that the type needs to support some more edge cases.

When using the search API, if no select is provided, all fields are returned. This is equivalent to `$select='*'`.  
When using the suggest API, if no select is provided, only the key field is returned.

This means that whatever type is returned by these calls, they need to:
1. Be aware of the amount of field selected.
2. Be aware of which field is the key field in the case of suggest.
3. Include `null` as a possible value for each field.
4. Probably ignore the "optionality" or `undefined` type of the original property

Reply:
@kawazoe good points. I've also reached the conclusion that just ignoring optionality in the model type is for the best. There is one case where we will need optional though: @dgetu and I have been discussing offline the case where the type of the select fields is not precisely known. For example, it should probably still be valid to provide `string[]` to the select (for example, if the selected fields are constructed programmatically), or at least `SelectFields<T>` (meaning we provided some set of fields, but don't know which one), and the result should be deeply partial, since we don't know enough in that case at compile time to refine the model.

The "standard result" seems like it should be a deeply nullable version of the model i.e., the result of calling the search API with no fields selected. So, our goal for search/get is:

- If `select` is known to be undefined (not clear exactly how we would test this, don't think `undefined extends Fields` will work), then we produce the standard result.
- If `select` is not precisely known (`SelectFields<T> extends Fields` or maybe `string[] extends Fields`), we produce a deeply _partial_ standard result. This is the most generically correct (widest type) thing that we can produce since it's correct for any combination of fields.
- If `Fields extends SelectFields<T>` produce the deeply picked standard result based on `Fields`.
- otherwise, should be some kind of assignability error and unreachable.

It's pretty difficult for us to know which field is the key in type space without forcing all users to add some kind of marker type to declare which field is the key. I don't know if we really want to get into the business of a type-space DSL for OData at present.

Maybe a happy medium is like what we do in the example `Hotel` interface now, where all the fields are `| null` except the key. So instead of us trying to do any fragile type calculus, we can declare which fields are keys by omitting `| null` from their declaration in the model interface. It's not ideal to have to put `| null` manually on every single field that _isn't_ the key, but it might be the best thing we can do unless/until TypeScript supports something like metadata decorators for types. With that assumption we could produce only the non-nullable fields of T in the suggest API. If the interface is well constructed that should only be the key field.

On the search service's metadata, I just don't know anything about it myself, so I'll leave that to @dgetu 😅. Probably best to make that a separate issue/PR if we're omitting some result data that should be produced in the results.

Reply:
@witemple-msft Personally, if marking the key field with a special type is not an option, I would prefer if the suggest endpoint would fallback to a deep partial type instead. Having to add `| null` to every field is very error prone and verbose. I also remember from before the rebrand to cognitive search that, as weird and unintuitive as it seams, you could actually index documents with null keys. Plus, I think it makes sense that the client doesn't assume what the API will provide when nothing explicit is requested. The API could very well change its defaults and break the types if they are too precise.

It would leave us with the following scenarios:

1. search without select => generate a type with all fields set to `TField | null`
2. search with select => generate a type with selected fields set to `TField | null`
3. search with vague select => generate a type with all fields set to `Optional<TField | null>`
4. suggest without select => generate a type with all fields set to `Optional<TField | null>`
5. suggest with select => generate a type with selected fields set to `TField | null`
6. suggest with a vague select => generate a type with all fields set to `Optional<TField | null>`

It feels like these 3 outcomes would cover every scenarios, and they are all fairly easy to detect while taking care of the null part automatically.

Reply:
> Having to add | null to every field is very error prone and verbose.
I agree. This latest revision matches the behavior you describe, but it breaks for users who define a non-partial non-nullable field in their model when they get a document and try to reuse it. I'm inclined to accept the widest return type possible as the parameter type for e.g. `uploadDocuments` and find some way to narrow it to the user-provided type in a non-breaking way.

Reply:
@dgetu Oh right, I was only considering querying, but it is true that a user should be able to upload a document, and then get it back with the same type without getting nulls on every fields. They already ensured that fields cannot be null when uploading it so it makes sense that this should mirrored when querying.

Reply:
I've tried a few things out, and what I've converged on so far is to assume that any non-nullable (string) fields in the user-defined model are the key field. Suggest will narrow its result to include only those fields by default. If the user chooses not to mark their non-key fields nullable/optional explicitly, it's up to the user to ensure that none of their queries result in a `null` or `undefined`/missing field being deserialized where it shouldn't be.